### DISCUSSION TITLE
Make programmatically added annotations visible in browser

### DISCUSSION THREAD

**Discussion by Excubitor** [mozilla/pdf.js#17649] [Q&A]
is there currently a way to make annotations manually added to the annotationstorage visible in the pdfjs viewer?

I can’t quite grasp how to reinitialize the page so that annotation in the storage gets rendered on screen.

**Comment by calixteman** [comment]
The  storage is just used to add the new properties which will be used when the pdf is printed/saved.
Originally it has been designed for forms, for example, to add a value entered by a user in a field, and then when the pdf is printed/saved the new value must be taken into account.
After it has been extended to support adding new annotations: the storage contains some serialized elements which can be deserialized (for example when on copy/paste we serialize/deserialize the annotation); for example you can have a look on:
https://github.com/mozilla/pdf.js/blob/master/src/display/editor/stamp.js#L515-L594

**Reply by gnarf2** [reply]
How likely is it that there will be an official way in the near future to store annotations/the annotation storage separately, for example to LocalStorage (and also load from there)? That would be really useful for a current project I'm working on. I'd implement something hacky myself, but I don't know if it would be worthwhile in the long run.

**Comment by lasseborly** [comment]
Yeah you need to handle that yourself. I'm just iterating through all inputs, textareas etc. and updating the value that way, so that it's the same as in the annotationstorage.

**Comment by Osiris-Team** [comment]
@calixteman I am trying to store annotations into localStorage so that I can keep them between site reloads, how would I do that?

**Reply by lasseborly** [reply]
There is no functionality for that built in. You need to manage that yourself and manually re-populate the input, textareas and whatnot. 

I am just gonna share my full implementation of this for others too see. Try to look past the specificities.
I do not have time to make an easier to digest example.

```javascript
/** @type {boolean} */
const userId = JSON.parse(document.getElementById("user-id").textContent);

/**
 * @typedef {Object} Character
 * @property {number} id
 * @property {string} character_name
 * @property {string} character_sheet
 * @property {Object} fields
 * @property {number} user
 */

/** @type {Character} */
const character = JSON.parse(document.getElementById("character").textContent);

/**
 * Most editing should only be enabled if the user looking
 * at the character is the one who owns it.
 */
const userOwned = character.user === userId;

// PDFViewerApplication.pdfDocument.annotationStorage
let storage;

/**
 * We need to make sure that we do not load the field values before
 * all pages have finsihed loading their annotation editor layers (inputs etc.)
 */
let editorLayersLoaded = 0;

/**
 * We store the id but need to get the DOM node from time to time.
 * @param {string} id
 * @returns
 */
function getFieldNode(id) {
  return document.querySelector(`[data-element-id="${id}"]`);
}

function setFields(fields) {
  for (const field in fields) {
    id = fields[field].id;
    node = getFieldNode(id);
    value = fields[field].value;

    if (node.type !== "checkbox") {
      node.value = value;
      storage.setValue(id, { value: value });
    }
    if (node.type === "checkbox") {
      node.checked = Boolean(value);
      storage.setValue(id, { value: Boolean(value) });
    }
  }
}

document.addEventListener("webviewerloaded", () => {
  PDFViewerApplicationOptions.set("disablePreferences", true);
  PDFViewerApplicationOptions.set("enableScripting", false);
  PDFViewerApplicationOptions.set("isEvalSupported", false);

  PDFViewerApplicationOptions.set(
    "defaultUrl",
    `/static/sheets/${character.character_sheet}.pdf`
  );

  PDFViewerApplication.initializedPromise.then(() => {
    PDFViewerApplication.eventBus.on("metadataloaded", () => {
      PDFViewerApplication.setTitle(character.character_name);

      // Sets the name of the downloaded file
      const slug = location.pathname
        .split("/")
        .filter((path) => path)
        .slice(-1)[0];
      PDFViewerApplication._contentDispositionFilename = `${slug}.pdf`;
    });

    PDFViewerApplication.eventBus.on(
      "annotationeditorlayerrendered",
      async () => {
        editorLayersLoaded += 1;
        if (editorLayersLoaded !== PDFViewerApplication.pagesCount) return;

        storage = PDFViewerApplication.pdfDocument.annotationStorage;
        setFields(character.fields);
        const field_objects =
          await PDFViewerApplication.pdfDocument.getFieldObjects();

        const fields_nodes = Object.values(field_objects).map((field) => {
          const node = getFieldNode(field[0].id);
          return node;
        });

        if (userOwned) {
          for (const node of fields_nodes) {
            if (node.value) {
              node.classList.add("annotation-layer-field__modified");
            }

            node.addEventListener("input", (event) => {
              if (event.target.value) {
                event.target.classList.add("annotation-layer-field__modified");
              } else {
                /**
                 * Keep the highlight for fields that aren't modified since it makes the
                 * fields easier to find.
                 */
                event.target.classList.remove(
                  "annotation-layer-field__modified"
                );
              }
            });

            node.setAttribute(
              "hx-post",
              `/update-character-field/${character.id}/`
            );
            if (node.type !== "checkbox") {
              node.setAttribute("hx-trigger", "keyup changed delay:250ms");
            }
            htmx.process(node);
          }
        } else {
          /**
           * If the user isn't the one who own's the character we do not want to
           * allow them to edit the fields. This is also checked in the BE to make
           * sure that the fields can't actually be changed.
           */
          for (const field of fields_nodes) {
            //there might be an official way of disabling this
            field.disabled = true;
            field.classList.add("annotation-layer-field__disabled");
          }
        }
      }
    );
  });
});

```

**Reply by Osiris-Team** [reply]
Thanks a lot will give it a try!

**Reply by Excubitor** [reply]
Trying to get it to work, but not making it so far.

**Reply by lasseborly** [reply]
Could you share the code you have @Excubitor ?

**Comment by Excubitor** [comment]
Not really a javascript guy. I'm trying to be able to reload annotations in a maui web view that loads the pdfjs official viewer. I can save and restore annotationstorage from the database (JSON.stringify/JSON.parse). I know I need to re-populate the inputs. I don't need to check the user. I allready know it is the correct user.

I have tried to use your code but haven't figured out what I should use for character and fields. I have a loaded a pdf and the restored data in the annotationstorage.

**Reply by Osiris-Team** [reply]
@lasseborly having the same issue, is there a way this can be turned into more like an API that provides an interface to add/remove/list/update annotations?

**Reply by Osiris-Team** [reply]
Im also not sure if this works with the latest version, especially `        PDFViewerApplication.eventBus.on(
            "annotationeditorlayerrendered",` doesnt seem to trigger when a new annotation is added (or even at all).


And `PDFViewerApplication.pdfDocument.getFieldObjects()` seems to return nothing.


The closest thing I found is listing the editor types: `pdfjsLib.AnnotationEditorLayer.#editorTypes.forEach((k,v)=>{console.log(k); console.log(v);})`

**Comment by Osiris-Team** [comment]
@Excubitor Update: Found a way to programmatically add fields/editors/annotations (however you want to call them):
https://github.com/Osiris-Team/pdf.js-utils

TODO for motivated people: make add function support drawings and images.

Tested on: PDF 110dd61fd57444010b1ab5ff38782f0f [1.4 pdfeTeX-1.21a / TeX] (PDF.js: 4.1.378 [a208d6bca])

**Reply by elia-c** [reply]
@Osiris-Team Do you have a solution to add an  svg image (Stamp editor) ?

**Reply by Osiris-Team** [reply]
@elia-c No, since I currently don't need it, however some tipps for anyone that wants to implement that:

- Instead of `pageEditorLayer.pasteEditor(pdfjsLib.AnnotationEditorType.FREETEXT` use stamp
- Create another function to set the default content.

Feel free to share your code.

**Reply by ademyalcin27** [reply]
I like your solution. thank you for solution. 
@Osiris-Team  do you have a solution for being able to do things like deleting the object you added to the pdf, relocating it, etc., but turning off editing? when you double click it turning to edit mode.

**Reply by Osiris-Team** [reply]
@ademyalcin27 nope :/ You should be able to change the position by setting the editor.x and editor.y variables and updating the width/height in the style.

**Reply by ademyalcin27** [reply]
@Osiris-Team Actually I would like to disable editing in Freetext mode. I want to add items to PDF like HTML. and I can not contentediable. whe you double click it turning contentediable true but I want to prevent it. There is defined double click event listener in freetext.js. I want to prevent it.


![Screenshot 2024-04-30 at 13 09 40](https://github.com/mozilla/pdf.js/assets/5026172/247ba48a-e861-477f-9c17-563251742922)

**Reply by Osiris-Team** [reply]
@ademyalcin27 idk maybe you can simply add the "disabled" attribute to that editor?

**Reply by Excubitor** [reply]
Getting back into this. The above seems to work great pasting text. I hope it is possible also for ink and highligts :)

**Reply by Excubitor** [reply]
How do you draw ink programatically? It doesn't seem to work to use ctx and canvas at it is drawn and then cleared.

**Reply by Osiris-Team** [reply]
I have no clue :/ If anyone finds a solution feel free to share, I'll need this too in the future.

**Reply by ademyalcin27** [reply]
I did some development about some adding items but different way. I will do the repository and video soon. I'll share it. I don't know when I can finish this.

**Reply by Osiris-Team** [reply]
@Excubitor this should help you out: https://stackoverflow.com/questions/22228552/serialize-canvas-content-to-arraybuffer-and-deserialize-again maybe even convert the binary to a base64 string.
This should work for ink, however maybe also for images, if they are in a canvas too I guess.

**Reply by berendberendsen** [reply]
@Osiris-Team have you figured out how to render annotations which come from a serialized file? I am able to store it and load it but it refuses to render. I have the feeling that I am doing it 'too late' and thus the render layer is already done, but I can't find the place where the storage is initialised

**Reply by Osiris-Team** [reply]
@berendberendsen the function addTextEditor(...) from the shown code should work.
In my case its a Java backend and the text fields as well as their html attributes are stored in a database.
So serialization and deserialization is done "manually".
Meaning i fetch the field and its HTML attributes from the db, and do something like this (pseudo code):
```
let f = addTextField(fieldValue);
for each key and value in fieldAttributes:
  f.div.setAttribute(key, value);
```

**Reply by berendberendsen** [reply]
Thanks @Osiris-Team, I am not sure if this will work for my case because I am trying to store and restore marker / highlighting annotations and the html generated for those is ... complex 😅 this is why I wanted to try if I could force pdfjs to see if it could draw the annotations. Curious to see if any of the others have managed this

**Reply by Osiris-Team** [reply]
@berendberendsen what do you mean by marker / highlighting?
There shouldn't be something like that yet, or do you mean the drawing / ink capabability?

**Reply by filip-ljubotina** [reply]
@Osiris-Team  I want to have a for loop that will programmatically add text annotation on all pages. When I do that with addTextEditor function it renders it for first two pages and then it catches an error "Cannot read properties of null (reading 'annotationEditorLayer')". Is such functionality even possible? I can see when debugging that only the page that I have rendered have AnnotationEditorLayer defined.

**Reply by berendberendsen** [reply]
@Osiris-Team there is actually functionality for it, you can enable a button with `enableHighlightEditor` in the viewer.js / viewer.mjs

**Reply by Osiris-Team** [reply]
@filip-ljubotina updated the code to wait for all pages to load their annotation editor layers, test it out now ;)
But if they really get defined only once the page is rendered this might be a problem.
A solution would then maybe be page.scrollIntoView() or something for each page that doesn't have the anootationEditor and finally do pages[0].scrollIntoView() once all are rendered.
Let me know if you found a solution!
Created this project with the current code here, so feel free to create a PR with your changes! https://github.com/Osiris-Team/pdf.js-utils

**Reply by Osiris-Team** [reply]
@berendberendsen wow nice, didn't know that. Take a look inside the addTextEditor(...), if a highlight is a separate editor object maybe something similar might be possible.

**Reply by berendberendsen** [reply]
@Osiris-Team that was my first thinking too, but then I started digging.. For the fun of it: try making a highlight and then inspect the html that is generated from it.. The text annotation is kind of nice and in one place

**Reply by Osiris-Team** [reply]
@berendberendsen meaning everything regarding the highlighting is directly in the text editors value? Then saving the value as html and repopulating it later should work, or doesnt it? If I remember correctly when I repopulated the text fields I had to encapsulate their values in `<div>`.

**Reply by berendberendsen** [reply]
@Osiris-Team from what I could see is that some of it refers to a div somewhere else again. Try highlighting a couple of paragraphs in a document. The html becomes quite messy quite quickly. It feels easier to somehow try to find the methods which actually draw these and use those based on what is stored in the annotationstorage but I can't seem to get that to work too. 

I will give it another try later, but if you find anything let me know!

**Reply by Osiris-Team** [reply]
@berendberendsen you might want to try using the profiler that chrome dev tools offers, press on record, do a highlighting and with luck find the related methods there.

**Reply by agss0** [reply]
@berendberendsen 
>  I have the feeling that I am doing it 'too late' and thus the render layer is already done, but I can't find the place where the storage is initialised

That was NOT a fun problem to figure out. I'm trying to save/load highlights from a YAML file when the PDF is loaded.

There are multiple ways to inject a highlight in a page. Highlights are stored in the `annotationStorage`, which is just hashmap of annotation objects. Adding naively an object to the storage like `annotationStorage.setValue(key, value)` would not work because the added object is a generic javascript object, meanwhile the "official" highlights are of the type `HighlightEditor`.

Making a `new HighlightEditor(...)` -ie. using the constructor of the class- wouldn't work because it needs the `annotationLayerEditor` object which only exists once the page is rendered (in the html sense).

Looking at the html, one can see, once the page is loaded:
```html
<div class="page" ...>
    <div class="canvasWrapper">...</div>
    <div class="textLayer">...</div>
    <div class="annotationLayer">...</div>
    <div class="annotationEditorLayer">...</div>
</div>
 ```
 Meanwhile an unloaded page looks like:
 ```html
 <div class="page" ...></div>
 ```
 
- the `canvasWrapper` is where the SVGs of the highlights (literaly the colored boxes) are added
- the `textLayer` contains the text of the page, think of it as the OCR of the page (if that makes sense)
- `annotationLayer` is where the data of the annotations can be accessed, eg. when you hover the highlight or click on it
- `annotationEditorLayer` is the interface to add/remove/edit/move/changeColor/etc the annotations

What it means is that you can't just "`add_to_storage([list_of_my_highlights])`" as soon as the PDF is ready, you need the page to exist before adding the highlight.

That said, there are two methods (afaik) to load the highlights:
1. Inject highlights as the page is loaded:
        There are some signals being triggered when things happens here and there. I've added the signal `"annotationEditorLayerRendered"` that is triggerd when the `<div class="annotationEditorLayer">` is added to the html. The highlights are added as you scroll the PDF, as the pages are done loading. In `pdfjs/web/pdf_page_view.js` at the end of the `async draw()` function: `this.eventBus.dispatch("annotationEditorLayerRendered", { pageNumber: this.id });`

2. Inject every highlights once when you load the PDF:
        Using the `PDFViewerApplication.pdfViewer.linkService.goToPage(page)` you can simulate scrolling the PDF: you jump to the nth page, the viewer loads it, you can inject etc.

Both those have a problem: the rendering of the page is async. There is a race condition happening on the EditorType. The call stack looks like this:

    `async render()`->...->`injectHighlight()`->`createAndAddNewEditor()`->`#createNewEditor()`->`#currentEditorType()`->`#uiManager.getMode()`

What is interesting is the uiManager's mode: `0=NONE`, `9=HIGHLIGHT`. My injecting function sets it to 9 before injecting the highlight, then sets it back to 0 when it's done.
PDF.js tries to pre-load the surrounding pages to reduce loading times; hence the signal `"annotationEditorLayerRendered"` happens twice (or more) and triggers the highlight injecting function in parallel, inducing a race condition on the `uiManager`'s mode.
Say we load the page 1:
    - `"annotationEditorLayerRendered"` detected (page 1)
    - `injectHighlight()` (page 1)
    - `uiManager.updateMode(9)` (page 1)
    - `"annotationEditorLayerRendered"` (page 2)
    - `injectHighlight()` (page 2)
    - `uiManager.updateMode(9)` (page 2)
    - injecting done, `uiManager.updateMode(0)` (page 1)
    - injecting done, `uiManager.updateMode(0)` (page 2)
The mode is set back to 0 before the `HighlightEditor` is added

My workaround is to hardcode "9" in `#createNewEditor()` instead of fetching the `uiManager`'s mode. We
can use the `"methodOfCreation"` property of the injected highlight's data:
```js
#createNewEditor(params) {
    let editorType;
    if (params.methodOfCreation === "artificial") {
        editorType = AnnotationEditorLayer.#editorTypes.get(AnnotationEditorType.HIGHLIGHT); // <- 9
    } else {
        editorType = this.#currentEditorType;
    }
    return editorType ? new editorType.prototype.constructor(params) : null;
}
```


Here's my injecting function:
```js
let injectedCount = 0;
function injectHighlight(pageNumber, data)
{
    const pageIndex = pageNumber - 1;
    const pageView = PDFViewerApplication.pdfViewer.getPageView(pageIndex);
    const annotationStorage = PDFViewerApplication.pdfDocument.annotationStorage;
    const annotationEditorLayer = pageView.annotationEditorLayer.annotationEditorLayer;

    if (!pageView)  console.warn("PageView not available");
    if (!annotationStorage) console.warn("Annotation storage unaccessible");
    if (!annotationEditorLayer) console.warn("Annotation Editor Layer unaccessible.\n The page has not been rendered entirely, some <div> are missing.");
    if (!pageView && !annotationStorage && !annotationEditorLayer) {
        return;
    }

    let highlight_id = "highlight_injected_" + injectedCount + "_p" + pageNumber;

    if (annotationStorage.has(highlight_id)) {
        console.warn(`[injectHighlight] Highlight ID ${highlight_id} already exists in storage before injection!`);
    }

    annotationEditorLayer.createAndAddNewEditor({
        x: 0,
        y: 0
    }, false, {
        id: highlight_id,
        methodOfCreation: "artificial",
        boxes: data.boxes,
        anchorNode: null,
        anchorOffset: 0,
        focusNode: {}, // div.textLayer.highlighting;
        focusOffset: 4,
        text: data.text,
    });

    injectedCount++;

    if (annotationStorage.has(highlight_id)) {
        console.info(`Highlight ${highlight_id} injected on page ${pageNumber}`);
    } else {
        console.error(`Could not add ${highlight_id} on page ${pageNumber} to storage for some reason`);
    }
}
```

And it is used like this:
```js
// inject the same highlight on every page once it is loaded

 const data = {
    boxes: [ { "x": 0.0713891806722689, "y": 0.574606357694593, "width": 0.18204638056394432, "height": 0.014854426619132501 } ],
 };

if (window.PDFViewerApplication && PDFViewerApplication.pagesCount > 0) {
    PDFViewerApplication.eventBus.on("annotationEditorLayerRendered", function(evt) {
        let page = evt.pageNumber;
        injectHighlight(page, data)
    }
}
```

- note that you need to check that the viewer and the PDF are ready
- the data is just some number but you can give more parameters like color or text

Hope this can help you.

**Reply by Osiris-Team** [reply]
Very cool @a-gss , what are the advantages over the current way of doing it ? Feel free to create a PR.

**Reply by agss0** [reply]
what do you mean "the current way" ?

**Reply by Osiris-Team** [reply]
see https://github.com/Osiris-Team/pdf.js-utils

**Reply by agss0** [reply]
Well as I understand it, you're polling each 100ms the annotationEditorLayer, while I trigger on its creation. This always works instantly, except for the two pages that are loaded at startup:
- the page 1 
- the last-time-you-left-the-pdf page
...and they sometime/sometime not trigger the "annotationEditorLayerRendered" event, thus not loading the highlights.

**Reply by agss0** [reply]
also, you're working around PDF.js, while I directly modify its source code to add some setters/getters and change some functions. So i can't really make a PR on your repo...

**Reply by Osiris-Team** [reply]
Ah I understand, when will this fix be available in the main repo?

**Reply by agss0** [reply]
Im not a PDF.js dev 🤣 
I just modified the source code of PDF.js according to my project needs

**Comment by h-sigma** [comment]
It would be quite useful if this was a dedicated API instead of a hacky method that may break anytime the library updates. It has immense usage for me, namely saving annotations separately for each user in my calibre-web instance.
